Kernel 6.12 compat: asm/unaligned.h shim (moved to linux/unaligned.h in 6.10, removed in 6.12), const ctl_table handlers, ip_route_output gained a scope argument, sysctl sentinels guarded, explicit linux/vmalloc.h include.

--- a/Makefile	2026-08-02 22:10:22.726186006 +0000
+++ b/Makefile	2026-08-02 22:16:05.471306012 +0000
@@ -1,3 +1,4 @@
+ccflags-y += -I$(obj)/compat
 ##########################################################################
 # Copyright (c) 2014-2016, 2018-2021, The Linux Foundation. All rights reserved.
 # Copyright (c) 2022-2023 Qualcomm Innovation Center, Inc. All rights reserved.
--- a/compat/asm/unaligned.h	1970-01-01 00:00:00.000000000 +0000
+++ b/compat/asm/unaligned.h	2026-08-02 22:16:05.471306012 +0000
@@ -0,0 +1,10 @@
+/* <asm/unaligned.h> was removed in kernel 6.12; contents moved to <linux/unaligned.h>. */
+#ifndef _ECM_COMPAT_ASM_UNALIGNED_H
+#define _ECM_COMPAT_ASM_UNALIGNED_H
+#include <linux/version.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0))
+#include <linux/unaligned.h>
+#else
+#include_next <asm/unaligned.h>
+#endif
+#endif
--- a/ecm_interface.c	2026-08-02 22:10:22.726186006 +0000
+++ b/ecm_interface.c	2026-08-02 22:16:05.461306012 +0000
@@ -1261,7 +1261,7 @@
 	 * it is using to communicate with that IP address.
 	 */
 	ECM_IP_ADDR_TO_NIN4_ADDR(be_addr, addr);
-	ecm_rt->rt.rtv4 = ip_route_output(&init_net, be_addr, 0, 0, 0);
+	ecm_rt->rt.rtv4 = ip_route_output(&init_net, be_addr, 0, 0, 0, 0);
 	if (IS_ERR(ecm_rt->rt.rtv4)) {
 		DEBUG_TRACE("No output route to: %pI4n\n", &be_addr);
 		return false;
@@ -1455,7 +1455,7 @@
 	__be32 ipv4_addr;
 
 	ECM_IP_ADDR_TO_NIN4_ADDR(ipv4_addr, addr);
-	rt = ip_route_output(&init_net, ipv4_addr, 0, 0, 0);
+	rt = ip_route_output(&init_net, ipv4_addr, 0, 0, 0, 0);
 	if (IS_ERR(rt)) {
 		return NULL;
 	}
@@ -8208,7 +8208,7 @@
  * ecm_interface_igs_enabled_handler()
  *	IGS enabled check sysctl node handler.
  */
-static int ecm_interface_igs_enabled_handler(struct ctl_table *ctl, int write, void __user *buffer,
+static int ecm_interface_igs_enabled_handler(const struct ctl_table *ctl, int write, void *buffer,
 		 size_t *lenp, loff_t *ppos)
 {
 	int ret;
@@ -8249,7 +8249,7 @@
  * ecm_interface_src_check_handler()
  *	Source interface check sysctl node handler.
  */
-static int ecm_interface_src_check_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int ecm_interface_src_check_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 	int current_value;
@@ -8303,7 +8303,9 @@
 		.proc_handler		= &ecm_interface_igs_enabled_handler,
 	},
 #endif
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 12, 0))
 	{ }
+#endif
 };
 
 #ifdef ECM_INTERFACE_IPSEC_GLUE_LAYER_SUPPORT_ENABLE
--- a/examples/ecm_ae_select.c	2026-08-02 22:10:22.721186006 +0000
+++ b/examples/ecm_ae_select.c	2026-08-02 22:16:05.471306012 +0000
@@ -52,7 +52,7 @@
  * ecm_ae_select_fallback_enable_handler()
  *	Fallback sysctl handler
  */
-int ecm_ae_select_fallback_enable_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_ae_select_fallback_enable_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 
@@ -82,7 +82,7 @@
  * ecm_ae_select_fallback_enable_handler()
  *	Fallback sysctl handler
  */
-int ecm_ae_test_select_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_ae_test_select_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 
@@ -131,7 +131,9 @@
 		.mode		= 0644,
 		.proc_handler	= &ecm_ae_test_select_handler,
 	},
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 12, 0))
 	{}
+#endif
 };
 
 /*
--- a/frontends/ecm_front_end_common.c	2026-08-02 22:10:22.723186006 +0000
+++ b/frontends/ecm_front_end_common.c	2026-08-02 22:24:50.767332005 +0000
@@ -18,6 +18,7 @@
  */
 
 #include <linux/version.h>
+#include <linux/vmalloc.h>
 #include <linux/module.h>
 #include <linux/of.h>
 #include <linux/debugfs.h>
@@ -835,7 +836,7 @@
  * ecm_front_end_ppe_fse_enable_limit_handler()
  *	Sysctl to enable/disable FSE programming through PPE.
  */
-int ecm_front_end_ppe_fse_enable_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_front_end_ppe_fse_enable_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 
@@ -870,7 +871,7 @@
  * ecm_front_end_db_conn_limit_handler()
  *	Database connection limit sysctl node handler.
  */
-int ecm_front_end_db_conn_limit_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_front_end_db_conn_limit_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 	int current_value;
@@ -1090,7 +1091,7 @@
  * ecm_front_end_udp_denied_ports_handler()
  *	Proc handler function for UDP denied ports read/write operation.
  */
-static int ecm_front_end_udp_denied_ports_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int ecm_front_end_udp_denied_ports_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	/*
 	 * Usage:
@@ -1110,7 +1111,7 @@
  * ecm_front_end_tcp_denied_ports_handler()
  *	Proc handler function for TCP denied ports read/write operation.
  */
-static int ecm_front_end_tcp_denied_ports_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int ecm_front_end_tcp_denied_ports_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	/*
 	 * Usage:
@@ -1157,7 +1158,9 @@
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_tcp_denied_ports_handler,
 	},
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 12, 0))
 	{}
+#endif
 };
 
 /*
--- a/frontends/sfe/ecm_sfe_common.c	2026-08-02 22:10:22.724186006 +0000
+++ b/frontends/sfe/ecm_sfe_common.c	2026-08-02 22:16:05.466306012 +0000
@@ -298,7 +298,7 @@
  * ecm_sfe_fast_xmit_enable_handler()
  *	Fast transmit sysctl node handler.
  */
-int ecm_sfe_fast_xmit_enable_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_sfe_fast_xmit_enable_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 
@@ -322,7 +322,7 @@
  * ecm_sfe_fse_enable_handler()
  *	Sysctl to enable/disable FSE programming through ECM SFE frontend.
  */
-int ecm_sfe_fse_enable_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+int ecm_sfe_fse_enable_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret;
 	int current_val;
@@ -416,7 +416,9 @@
 		.mode           = 0644,
 		.proc_handler   = &ecm_sfe_fse_enable_handler,
 	},
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 12, 0))
 	{}
+#endif
 };
 
 /*
